### ISSUE TITLE
[linux] Minimum supported GCC/Clang version

### ISSUE THREAD

**Issue by antiagainst** [google/DirectXShaderCompiler#196]
We need to set a minimum supported GCC/Clang version and have the CIs testing them.

I'm thinking about
* GCC: 5.3, released on 2015-12-04 (https://gcc.gnu.org/releases.html)
* Clang: 3.8, released on 2016-03-08 (http://releases.llvm.org/)

Why these two versions? My anchor is the previous LTS version of Ubuntu: 16.04. It ships with the above two versions.
* GCC: https://packages.ubuntu.com/xenial/gcc
* Clang: https://packages.ubuntu.com/xenial/clang

@ehsannas @pow2clk: Any objections/thoughts?

**Comment by pow2clk** [comment]
I think it bears mentioning that while C++11 support is greatly improved in GCC5 over GCC4, it's still not enabled by default. That means a flag would have to be set. I'll see if I have more success installing GCC5 than I did GCC4 and see how it does with the build.

**Comment by pow2clk** [comment]
I was more successful in getting gcc5. However it was not more successful at processing  enum values referenced using scope operators :: than GCC4. That's something we could conceivably fix, but it might be a lot of work to remove what is a valid C++11 functionality.  I don't know how pervasive it is.

**Comment by antiagainst** [comment]
Regarding default C++ version, I think GCC [5 is `-std=gnu11`](https://gcc.gnu.org/gcc-5/changes.html) and [6 is `-std=gnu14`](https://gcc.gnu.org/gcc-6/changes.html)?

**Comment by antiagainst** [comment]
>  However it was not more successful at processing enum values referenced using scope operators :: than GCC4. That's something we could conceivably fix, but it might be a lot of work to remove what is a valid C++11 functionality. I don't know how pervasive it is.

Interesting. Thanks for the update! I don't want to create lots of work for you by requiring GCC 5. My initial thought was that presumably lots of people are still using Ubuntu 16.04 LTS, so it would be nice to give them a hassle-free way of compiling DXC. I think if we can support Clang 3.8, then that goal is somewhat satisfied; people will just need `apt install clang` and set `CMAKE_{C|CXX}_COMPILER`: No need to register a PPA or build the toolchain by themselves. So I think I'm Okay of pinpointing the minimal GCC version to be 6 for now and maybe consider lower it later.

**Comment by pow2clk** [comment]
Well. The documentation for GCC6 says that it increased the default C++ version from 98 to 14, which suggests they skipped 11. However, in my experience, it does not support C++14 features such as make_unique unless you provide a flag. GCC5 is supposed to support all the enum class features, but doesn't allow scoping. So there does seem to be something of a disconnect between release docs and reality on the ground there.

I agree that GCC5 seems like a reasonable minimum, but it's the features used in the code that complicate that. I think LLVM has mostly been built with clang, which makes sense for the most part, but I think it'll be helpful to not require linux users who don't have clang installed to use gcc if they want dxc. GCC is still overwhelmingly more common on linux installations than clang I think.

I'll keep this in mind and maybe poke at it a bit more. I'll let you know if anything I've said turns out to be mistaken or if another solution comes up.

**Comment by pow2clk** [comment]
I just made a little test program (tm). That suggests the problem is not as widespread as I'd thought. Instead it seems that the case in question in ScalarReplAggregatesHLSL.cpp was complicated by the HLSL code's peculiar proclivity to give types and variables the same names. Given that it works fine on GCC6 and doesn't on GCC5 or 4, it's likely that this is technically a bug in GCC that was fixed without a lot of fanfare in 6, but it still strikes me as weird and potentially confusing to readers.

The fix for that is't too onerous. I'll see where the build goes from here.

To summarize, you still need to specify --std=c++11 on GCC versions prior to 6, but the problem with scoped enums isn't that they don't work so much as there is a fairly contained bug in the implementation. 


Here's the little test program (tm) if you're curious:

`enum class Letters { A, B, C }`
`#if BUGGY`
`Letters;`
`#else`
`letters;`
`#endif`
`bool isA(enum Letters letter) {`
`  if (letter == Letters::A) return true; // This is where it will error if BUGGY is set`
`  return false;`
`}`

**Comment by pow2clk** [comment]
https://github.com/google/DirectXShaderCompiler/pull/197 fixes these issues. It includes a change I was working on earlier for GCC6 support but ultimately considered unnecessary. The bulk of the change is renaming variables that share a name with a type. Something GCC disdains and this code base seems to love.

**Comment by pow2clk** [comment]
I think this issue is resolved?

**Comment by ehsannas** [comment]
I don't think there's anything left to do. Once we create a doc file (or Wiki page) for the Linux compilation in the MS repo, we should mention the minimum version that is supported.
